# Airbnb no Rio de Janeiro - Limpeza de Dados

Os dados desse projeto foram retirados do site: [Inside Airbnb](http://insideairbnb.com/get-the-data.html), que disponibliza datasets de dados coletados do próprio site do Airbnb. 

Foi utilizado um dataset referente aos dados de Airbnb no Rio de Janeiro - RJ. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('data/listings.csv')

In [3]:
df.head()

id                         listing_url       scrape_id last_scraped  \
0  17878  https://www.airbnb.com/rooms/17878  20211026045247   2021-10-26   
1  24480  https://www.airbnb.com/rooms/24480  20211026045247   2021-10-26   
2  25026  https://www.airbnb.com/rooms/25026  20211026045247   2021-10-26   
3  35636  https://www.airbnb.com/rooms/35636  20211026045247   2021-10-27   
4  35764  https://www.airbnb.com/rooms/35764  20211026045247   2021-10-26   

                                                name  \
0  Very Nice 2Br in Copacabana w. balcony, fast WiFi   
1   Nice and cozy near Ipanema Beach, w/ home office   
2          Beautiful Modern Decorated Studio in Copa   
3                   Cosy flat close to Ipanema beach   
4     COPACABANA SEA BREEZE  -  RIO - 20 X Superhost   

                                         description  \
0  Discounts for long term stays. <br />- Large b...   
1  My studio is located in the best of Ipanema, t...   
2  Our apartment is a little gem, everyone loves ...   
3  This cosy apartment  is just a few steps away ...   
4  Our newly renovated studio is located in the b...   

                               neighborhood_overview  \
0  This is the one of the bests spots in Rio. Bec...   
1  The beach, the lagoon, Ipanema is a great loca...   
2  Copacabana is a lively neighborhood and the ap...   
3  The  apartment street is very quiet and safe ....   
4  Our guests will experience living with a local...   

                                         picture_url  host_id  \
0  https://a0.muscache.com/pictures/65320518/3069...    68997   
1  https://a0.muscache.com/pictures/11955612/b28e...    99249   
2  https://a0.muscache.com/pictures/3003965/68ebb...   102840   
3  https://a0.muscache.com/pictures/20009355/38b6...   153232   
4  https://a0.muscache.com/pictures/23782972/1d3e...   153691   

                                   host_url  ... review_scores_communication  \
0   https://www.airbnb.com/users/show/68997  ...                        4.90   
1   https://www.airbnb.com/users/show/99249  ...                        4.90   
2  https://www.airbnb.com/users/show/102840  ...                        4.90   
3  https://www.airbnb.com/users/show/153232  ...                        4.88   
4  https://www.airbnb.com/users/show/153691  ...                        4.94   

  review_scores_location review_scores_value license instant_bookable  \
0                   4.76                4.65     NaN                f   
1                   4.97                4.58     NaN                f   
2                   4.81                4.56     NaN                f   
3                   4.90                4.65     NaN                f   
4                   4.94                4.87     NaN                f   

  calculated_host_listings_count calculated_host_listings_count_entire_homes  \
0                              1                                           1   
1                              1                                           1   
2                              1                                           1   
3                              1                                           1   
4                              1                                           1   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              2.07  
1                                           0              0.85  
2                                           0              1.84  
3                                           0              1.90  
4                                           0              3.05  

[5 rows x 74 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23359 entries, 0 to 23358
Data columns (total 74 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            23359 non-null  int64  
 1   listing_url                                   23359 non-null  object 
 2   scrape_id                                     23359 non-null  int64  
 3   last_scraped                                  23359 non-null  object 
 4   name                                          23338 non-null  object 
 5   description                                   22137 non-null  object 
 6   neighborhood_overview                         12726 non-null  object 
 7   picture_url                                   23359 non-null  object 
 8   host_id                                       23359 non-null  int64  
 9   host_url                                      23359 non-null 

**Criação de um DataFrame somente com as colunas que serão utilizadas**

In [5]:
df = df[['id', 'name', 'host_id', 'neighbourhood_cleansed', 'latitude', 'longitude', 
         'property_type', 'room_type', 'accommodates', 'bathrooms_text', 'bedrooms', 
         'beds', 'price', 'minimum_nights', 'maximum_nights', 'number_of_reviews', 
         'reviews_per_month']]

In [6]:
df.head()

id                                               name  host_id  \
0  17878  Very Nice 2Br in Copacabana w. balcony, fast WiFi    68997   
1  24480   Nice and cozy near Ipanema Beach, w/ home office    99249   
2  25026          Beautiful Modern Decorated Studio in Copa   102840   
3  35636                   Cosy flat close to Ipanema beach   153232   
4  35764     COPACABANA SEA BREEZE  -  RIO - 20 X Superhost   153691   

  neighbourhood_cleansed  latitude  longitude               property_type  \
0             Copacabana -22.96599  -43.17940  Entire condominium (condo)   
1                Ipanema -22.98405  -43.20189          Entire rental unit   
2             Copacabana -22.97735  -43.19105          Entire rental unit   
3                Ipanema -22.98839  -43.19232          Entire rental unit   
4             Copacabana -22.98107  -43.19136                 Entire loft   

         room_type  accommodates bathrooms_text  bedrooms  beds    price  \
0  Entire home/apt             5         1 bath       2.0   2.0  $350.00   
1  Entire home/apt             2         1 bath       1.0   1.0  $296.00   
2  Entire home/apt             2         1 bath       1.0   1.0  $168.00   
3  Entire home/apt             2      1.5 baths       1.0   1.0  $278.00   
4  Entire home/apt             2      1.5 baths       1.0   1.0  $150.00   

   minimum_nights  maximum_nights  number_of_reviews  reviews_per_month  
0               5             180                266               2.07  
1               3              90                 85               0.85  
2               4              90                238               1.84  
3               2              89                181               1.90  
4               3              15                373               3.05

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23359 entries, 0 to 23358
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      23359 non-null  int64  
 1   name                    23338 non-null  object 
 2   host_id                 23359 non-null  int64  
 3   neighbourhood_cleansed  23359 non-null  object 
 4   latitude                23359 non-null  float64
 5   longitude               23359 non-null  float64
 6   property_type           23359 non-null  object 
 7   room_type               23359 non-null  object 
 8   accommodates            23359 non-null  int64  
 9   bathrooms_text          23291 non-null  object 
 10  bedrooms                21871 non-null  float64
 11  beds                    23167 non-null  float64
 12  price                   23359 non-null  object 
 13  minimum_nights          23359 non-null  int64  
 14  maximum_nights          23359 non-null

## Lidando com Dados Nulos

**Verificação de quais colunas possuem dados nulos (em ordem decrescente)**

In [8]:
df.isna().sum().sort_values(ascending=False)

reviews_per_month         8252
bedrooms                  1488
beds                       192
bathrooms_text              68
name                        21
number_of_reviews            0
maximum_nights               0
minimum_nights               0
price                        0
id                           0
room_type                    0
property_type                0
longitude                    0
latitude                     0
neighbourhood_cleansed       0
host_id                      0
accommodates                 0
dtype: int64

**Abordagem**

Como nenhuma coluna possui mais de 50% dos valores faltantes, não excluiremos nenhuma coluna. Com relação as linhas faltantes, ao invés de excluir linhas que não possuem informações, adotaremos uma abordagem de tentar preencher as linhas com uma suposição razoável, seja por técnicas estatísticas ou por um conhecimento do domínio do problema. 

**Coluna reviews_per_months**

A coluna *reviews_per_month* (avaliações por mês) tem 8252 linhas nulas. Isso representa aproximadamente 35% da quantidade total de dados. É uma quantidade muito expressiva para descartar todas essas linhas.

Vamos assumir que como não há uma quantidade de avaliações por mês, então não foi feita nenhuma avaliação para aquele Airbnb. O que significa que podemos preencher essas linhas com o valor 0.

In [9]:
df['reviews_per_month'].fillna(0, inplace=True)

In [10]:
df.isna().sum().sort_values(ascending=False)

bedrooms                  1488
beds                       192
bathrooms_text              68
name                        21
id                           0
number_of_reviews            0
maximum_nights               0
minimum_nights               0
price                        0
accommodates                 0
room_type                    0
property_type                0
longitude                    0
latitude                     0
neighbourhood_cleansed       0
host_id                      0
reviews_per_month            0
dtype: int64

**Coluna bedrooms**

A coluna *bedrooms* (quartos) possui 1488 linhas nulas. Isso representa aproximadamente 6% dos dados. Apesar de ser uma pequena porcentagem, íremos preencher essas linhas com a mediana dos demais valores existentes nessa coluna. 

Foi escolhido o uso da mediana porque apesar de dos valores nessa coluna serem do tipo float64, podemos observar que todos os números são inteiros. Caso usassemos a média, obteríamos 1.7, um número não inteiro, e como não é possível existir 1.7 quartos, usamos a mediana para obtermos um valor inteiro. 

In [11]:
# Verificando os valores únicos da coluna
df['bedrooms'].unique()

array([ 2.,  1.,  6.,  4.,  3., nan,  7., 10.,  5.,  9., 15., 20.,  8.,
       17., 12., 11., 13., 30., 14., 47., 18.])

In [12]:
df['bedrooms'].median()

1.0

In [13]:
df['bedrooms'].fillna((df['bedrooms'].median()), inplace=True)

In [14]:
df.isna().sum().sort_values(ascending=False)

beds                      192
bathrooms_text             68
name                       21
id                          0
number_of_reviews           0
maximum_nights              0
minimum_nights              0
price                       0
bedrooms                    0
accommodates                0
room_type                   0
property_type               0
longitude                   0
latitude                    0
neighbourhood_cleansed      0
host_id                     0
reviews_per_month           0
dtype: int64

**Coluna beds**

A coluna beds (camas) possui 192 valores nulos, o que representa aproximadamente 0.7% dos dados totais. Usando a mesma lógica de preenchimento da coluna bedrooms usaremos a mediana dos valores existentes.

In [15]:
df['beds'].unique()

array([ 2.,  1.,  7.,  5.,  3., 10.,  6.,  4.,  8.,  0.,  9., 16., 50.,
       nan, 14., 13., 17., 12., 15., 11., 18., 91., 24., 35., 30., 41.,
       38., 20., 22., 27., 33., 21., 26., 25., 23., 43., 52., 29., 31.])

In [16]:
df['beds'].median()

2.0

In [17]:
df['beds'].fillna((df['beds'].median()), inplace=True)

In [18]:
df.isna().sum().sort_values(ascending=False)

bathrooms_text            68
name                      21
id                         0
number_of_reviews          0
maximum_nights             0
minimum_nights             0
price                      0
beds                       0
bedrooms                   0
accommodates               0
room_type                  0
property_type              0
longitude                  0
latitude                   0
neighbourhood_cleansed     0
host_id                    0
reviews_per_month          0
dtype: int64

**Coluna bathrooms_text**

A coluna *bathrooms_text* (textos sobre banheiros) tem 68 linhas nulas. Isso representa menos de 1% do total de dados. Entretanto, ao invés de excluir essas linhas do dataset, podemos preenchê-las com a moda, ou seja, o valor mais frequente dessa coluna. Usaremos a moda porque os valores são strings, então vamos tratá-los como dados categóricos nesse primeiro momento.

In [19]:
df['bathrooms_text'].unique()

array(['1 bath', '1.5 baths', '15 baths', '1 private bath', '3 baths',
       '7 baths', '4 baths', '2.5 baths', '2 baths', '5 shared baths',
       nan, '1 shared bath', '4.5 baths', '2 shared baths',
       '1.5 shared baths', '3.5 baths', '0 shared baths', '5 baths',
       '0 baths', '2.5 shared baths', '6 baths', 'Shared half-bath',
       '10 baths', '8 baths', '3 shared baths', '6.5 baths', '5.5 baths',
       '4 shared baths', '7.5 baths', 'Half-bath', '3.5 shared baths',
       '7 shared baths', '11.5 baths', '20 shared baths', '20 baths',
       '6.5 shared baths', '11 baths', '12 baths', '15.5 baths',
       '8.5 baths', '6 shared baths', 'Private half-bath', '10.5 baths',
       '13 baths', '9 baths', '4.5 shared baths', '18 baths',
       '8.5 shared baths', '50 baths', '9.5 baths', '9 shared baths',
       '8 shared baths', '5.5 shared baths'], dtype=object)

In [20]:
df['bathrooms_text'].value_counts().index[0]

'1 bath'

In [21]:
df['bathrooms_text'].fillna(df['bathrooms_text'].value_counts().index[0], inplace=True)

In [22]:
df.isna().sum().sort_values(ascending=False)

name                      21
id                         0
bathrooms_text             0
number_of_reviews          0
maximum_nights             0
minimum_nights             0
price                      0
beds                       0
bedrooms                   0
accommodates               0
room_type                  0
property_type              0
longitude                  0
latitude                   0
neighbourhood_cleansed     0
host_id                    0
reviews_per_month          0
dtype: int64

**Coluna name**

A coluna name possui apenas 21 linhas com dados nulos. Apesar de ser uma coluna que não será útil para o modelo de machine learning, essa pode ser uma coluna interessante para algumas visualiações. Se essa coluna possuir valores nulos, pode ser que ocorram erros nos gráficos a utilizarem. Por isso, vamos simplesmente preencher esses nomes nulos com a string 'Name not informed'.

In [23]:
df['name'].fillna('Name not informed', inplace=True)

In [24]:
df.isna().sum().sort_values(ascending=False)

id                        0
bathrooms_text            0
number_of_reviews         0
maximum_nights            0
minimum_nights            0
price                     0
beds                      0
bedrooms                  0
accommodates              0
name                      0
room_type                 0
property_type             0
longitude                 0
latitude                  0
neighbourhood_cleansed    0
host_id                   0
reviews_per_month         0
dtype: int64

## Transformando strings em valores numéricos

A coluna price (preço) está em forma de string. Vamos remover o \\$ e transformar o valor em um dado numérico. Devemos notar que além do sinal de \\$, devemos tomar cuidado com alguns valores que possuem , (vírgula), pois se a vírgula não for removida ocorrerá um erro na conversão de string para float.

In [25]:
df['price'].unique()

array(['$350.00', '$296.00', '$168.00', ..., '$1,083.00', '$2,456.00',
       '$408.00'], dtype=object)

In [26]:
def convert_price_to_numeric(price):
    return float(price[1:].replace(',', ''))

In [27]:
df['price'] = df['price'].apply(convert_price_to_numeric)

## Criando novas features com a informação sobre banheiro

In [28]:
df['bathrooms_text'].unique()

array(['1 bath', '1.5 baths', '15 baths', '1 private bath', '3 baths',
       '7 baths', '4 baths', '2.5 baths', '2 baths', '5 shared baths',
       '1 shared bath', '4.5 baths', '2 shared baths', '1.5 shared baths',
       '3.5 baths', '0 shared baths', '5 baths', '0 baths',
       '2.5 shared baths', '6 baths', 'Shared half-bath', '10 baths',
       '8 baths', '3 shared baths', '6.5 baths', '5.5 baths',
       '4 shared baths', '7.5 baths', 'Half-bath', '3.5 shared baths',
       '7 shared baths', '11.5 baths', '20 shared baths', '20 baths',
       '6.5 shared baths', '11 baths', '12 baths', '15.5 baths',
       '8.5 baths', '6 shared baths', 'Private half-bath', '10.5 baths',
       '13 baths', '9 baths', '4.5 shared baths', '18 baths',
       '8.5 shared baths', '50 baths', '9.5 baths', '9 shared baths',
       '8 shared baths', '5.5 shared baths'], dtype=object)

**Transformando linhas que não possuem valor numérico para a quantidade de banheiros**

As linhas que possuem 'Shared half-bath', 'Half-bath' ou 'Private half-bath' na coluna de bathrooms_text não têm um valor numérico para a quantidade de banheiros. Como são half-baths (metade de um banheiro), isso significa que são banheiros sem chuveiro e podemos assumir que isso está representado nos dados pelo valor 0.5.

In [29]:
half_baths = {'Shared half-bath': '0.5 shared half-bath', 
              'Half-bath': '0.5 half-bath',
              'Private half-bath': '0.5 half-bath'}

In [30]:
df['bathrooms_text'].replace(half_baths, inplace=True)

In [31]:
df['bathrooms_text'].unique()

array(['1 bath', '1.5 baths', '15 baths', '1 private bath', '3 baths',
       '7 baths', '4 baths', '2.5 baths', '2 baths', '5 shared baths',
       '1 shared bath', '4.5 baths', '2 shared baths', '1.5 shared baths',
       '3.5 baths', '0 shared baths', '5 baths', '0 baths',
       '2.5 shared baths', '6 baths', '0.5 shared half-bath', '10 baths',
       '8 baths', '3 shared baths', '6.5 baths', '5.5 baths',
       '4 shared baths', '7.5 baths', '0.5 half-bath', '3.5 shared baths',
       '7 shared baths', '11.5 baths', '20 shared baths', '20 baths',
       '6.5 shared baths', '11 baths', '12 baths', '15.5 baths',
       '8.5 baths', '6 shared baths', '10.5 baths', '13 baths', '9 baths',
       '4.5 shared baths', '18 baths', '8.5 shared baths', '50 baths',
       '9.5 baths', '9 shared baths', '8 shared baths',
       '5.5 shared baths'], dtype=object)

**Criando coluna com o valor numérico da quantidade de banheiro**

In [32]:
def get_bath_qnt(bath_text):
    bath_qnt = float(bath_text.split()[0])
    return bath_qnt

In [33]:
df['bathrooms_number'] = df['bathrooms_text'].apply(get_bath_qnt)

**O banheiro é compartilhado? (Bathroom is shared?)**

Criando nova feature que diz se o(s) banheiro(s) são compartilhados ou não.

In [34]:
def is_shared(bath):
    if bath.lower().find('shared') < 0:
        return 0
    else:
        return 1

In [35]:
df['bathroom_is_shared'] = df['bathrooms_text'].apply(is_shared)

**Tipo de banheiro**

In [36]:
def get_bath_type(bath):
    bath_list = bath.split()[1:]
    bath_type = ''.join([f'{bath} ' for bath in bath_list])
    
    bath_type = bath_type.replace('private ', '')
    
    if bath_type[-2] == 's':
        return bath_type[:-2]
    else:
        return bath_type[:-1]

In [37]:
df['bathrooms_type'] = df['bathrooms_text'].apply(get_bath_type)

In [38]:
df['bathrooms_type'].unique()

array(['bath', 'shared bath', 'shared half-bath', 'half-bath'],
      dtype=object)

## Salvando o dataset limpo

In [39]:
df.to_csv('data/cleaned_listing.csv')